[GStreamer] fast/mediastream/stream-switch.html is frequently flaky timing out since 292066@main

philn opened the pull request:
#### be4b0a445375c6941b751ed2ded71d3f17d6dc44
<pre>
[GStreamer] fast/mediastream/stream-switch.html is frequently flaky timing out since 292066@main
<a href="https://bugs.webkit.org/show_bug.cgi?id=292842">https://bugs.webkit.org/show_bug.cgi?id=292842</a>

Reviewed by Xabier Rodriguez-Calvar.

The test was timing out due to a deadlock in the WebProcess, after signaling EOS the element was
posting a new stream collection while the EOS event was still traveling towards the sinks. Now when
EOS is signalled from outgoing WebRTC sources and MediaRecorder we at least wait the event has been
pushed downstream before emitting the new collection and cleaning-up resources. There are still
issues in the MediaRecorder which sometimes keep waiting for EOS on its sink forever, for the time
being a watch-dog works around this issue which is tracked in bug 293124.

Canonical link: <a href="https://commits.webkit.org/295203@main">https://commits.webkit.org/295203@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a345d7ffec5875c05e8e38b321af443ec565e337

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104254 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23958 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14342 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109454 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54922 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106294 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24328 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32504 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79181 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107260 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18904 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94082 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59509 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18711 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12137 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54282 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88413 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12198 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111836 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31410 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23152 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88202 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31775 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90302 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87865 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22397 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32757 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10519 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25877 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31339 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36652 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31133 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34469 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32693 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.cpp:
```diff
@@ -609,37 +667,39 @@ void RealtimeOutgoingMediaSourceGStreamer::stopUpdatingStats()
-    m_sender.clear();
-    m_webrtcSinkPad.clear();
-    m_parameters.reset();
+        gst_element_set_locked_state(m_bin.get(), TRUE);
+        gst_element_set_state(m_bin.get(), GST_STATE_NULL);
+        if (auto pipeline = adoptGRef(gst_element_get_parent(m_bin.get())))
+            gst_bin_remove(GST_BIN_CAST(pipeline.get()), m_bin.get());
+        gst_element_set_locked_state(m_bin.get(), FALSE);
```
I wonder if we should create an external locker for this.

philn replied:
OK I can follow-up on that.

I'm not sure where you think the commit msg has a parse error. Grammar looks OK :)

calvaris reviewed (approved):
I can't parse this sentence in the changelog:

> Now when
EOS is signalled from outgoing WebRTC sources and MediaRecorder we at least wait the event has been
pushed downstream before emitting the new collection and cleaning-up resources

philn committed:
[GStreamer] fast/mediastream/stream-switch.html is frequently flaky timing out since 292066@main
https://bugs.webkit.org/show_bug.cgi?id=292842

Reviewed by Xabier Rodriguez-Calvar.

The test was timing out due to a deadlock in the WebProcess, after signaling EOS the element was
posting a new stream collection while the EOS event was still traveling towards the sinks. Now when
EOS is signalled from outgoing WebRTC sources and MediaRecorder we at least wait the event has been
pushed downstream before emitting the new collection and cleaning-up resources. There are still
issues in the MediaRecorder which sometimes keep waiting for EOS on its sink forever, for the time
being a watch-dog works around this issue which is tracked in bug 293124.

Canonical link: https://commits.webkit.org/295203@main
